### PR TITLE
Fix webspace name with special characters in it

### LINKED ISSUE #6706: Webspace name doesn't properly encode all characters
| Q | A
| --- | ---
| Sulu Version | 2.5.0
| PHP Version | 8.1.7
| DB Version | mysql-server:8.0
| Browser Version | 101.0.1

# Actual Behavior

When I configure a webspace name to contain special characters. Some of the special characters like the umlauts work but things like the `'` do not encode correctly.
webspace

# Expected Behavior
When I configure a webspace in the XML file I expect the name to be displayed correctly.

# Steps to Reproduce
Create an empty sulu project and configure the webspace's name like so.
```xml
<webspace>
    <name>Günter's Homepage</name>
    <key>homepage</key>
</webspace>
```

# Possible Solutions
I don't know if the problem is in the storing of the data but when I checked it it looks like the data was incorrect when loaded from the database so it is probably not a frontend issue.

### PR DESCRIPTION
**Pull request by mamazu**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #6706
| Related issues/PRs | -
| License | MIT
| Documentation PR | -

#### What's in this PR?

This PR will fix the generated code for the webspace collection cache.

#### Why?

If you configure a name in the XML file you also want to have it be displayed the same on the website. For an example see #6706

You can't just use the raw filter because this will create an issue with the surrounding code resulting in something like this:
```php
$webspace->setName('Tim's Webspace');
```

#### Example Usage

Just like before and now you can also have apostrophies in your webspace name.

#### To Do

- [x] Create a documentation PR (not needed)
- [x] Add breaking changes to UPGRADE.md (not needed)

### CODE REVIEW

**Comment by mamazu** [comment]
Codestyle is broken but it can't be my class since this isn't even a proper class. :D

File: src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig [hunk 1]
```diff
@@ -32,7 +32,7 @@ class {{ cache_class }} extends {{ base_class }}
         // new webspace
         $webspace = new Webspace();
         $webspace->setKey('{{ webspace.key }}');
-        $webspace->setName('{{ webspace.name }}');
+        $webspace->setName(html_entity_decode('{{ webspace.name }}'));
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig]: We should write the properties already correctly into this cache file.

The problem here currently is the "Twig" autoescaping. Which I think we should disable and go with a "json_encode" solution here:

https://twigfiddle.com/d3hoih

https://twigfiddle.com/d3hoih

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
Good catch. Some small comment.

File: src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig (hunk 1 above)
Comment (mamazu) [reply src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig]: Yes this makes it easier. I have changed it to JSON now.

File: src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig [hunk 2]
```diff
@@ -32,7 +32,7 @@ class {{ cache_class }} extends {{ base_class }}
         // new webspace
         $webspace = new Webspace();
         $webspace->setKey('{{ webspace.key }}');
-        $webspace->setName(html_entity_decode('{{ webspace.name }}'));
+        $webspace->setName(json_decode({{ webspace.name|json_encode }}));
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig]: json_decode is not required here:

```suggestion
        $webspace->setName({{ webspace.name|json_encode|raw }});
```

Comment (mamazu) [reply src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig]: You're right. :D

File: src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig [hunk 3]
```diff
@@ -32,7 +32,7 @@ class {{ cache_class }} extends {{ base_class }}
         // new webspace
         $webspace = new Webspace();
         $webspace->setKey('{{ webspace.key }}');
-        $webspace->setName('{{ webspace.name }}');
+        $webspace->setName({{ webspace.name|json_encode }});
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Resources/skeleton/WebspaceCollectionClass.php.twig]:
```suggestion
        $webspace->setName({{ webspace.name|json_encode|raw }});
```

raw is required or as written in the comment autoescape false ;)

**Comment by alexander-schranz** [comment]
Can you add a testcase for your case?

**Comment by mamazu** [comment]
Where would be the best way to put it?

**Comment by mamazu** [comment]
And what about the other code that is failing the pipeline?

**Comment by alexander-schranz** [comment]
We sadly did not yet found out why the PHP 8.0 tests in 90% run into a timeout but work without problems locally. Funfact it seems always be that there is atleast one CI whatever it is which just like to go into timeout instead of running finish a job. Still there are PR where all runs without any problems.

The `Test application / PHP Lint (pull_request)` can be ignored. If it doesn't show anything related to your pull request. We mostly get new versions of phpstan or php-cs-fixer which can make the CI fail. But that is a job we are doing and contributers shoud focus on the PR itself :)

**Comment by alexander-schranz** [comment]
For the test I would create a unit test for the `PhpWebspaceCollectionDumper` class.

Somewhere here: https://github.com/sulu/sulu/tree/2.4/src/Sulu/Component/Webspace/Tests/Unit/Manager

By creating a `Dumper` namespace.

You can create the given `WebspaceCollection` by creating one via:

```php
$webspaceCollection = new WebspaceCollection();
// and add then different webspace
$webspace = new Webspace();
$webspaceCollection->add(...);
```

Call the dumper then and after it you `require` the generated collection and call the getter to test if you get the correct name.

**Comment by alexander-schranz** [comment]
@mamazu looks good. Can we rebase this branch to the `2.4` and set the target branch to 2.4? 

Copy & Paste rebase script:

```bash
git clone git@github.com:mamazu/sulu.git
cd sulu
git checkout webspace_name_bug

git remote add sulu git@github.com:sulu/sulu.git
git fetch sulu

git rebase -i sulu/2.5 --onto sulu/2.4
# close shown file

git push origin webspace_name_bug --force # force required in this case
```

**Comment by mamazu** [comment]
Should be done.

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        $this->setWebspaces($webspaceRefs);
+        $this->setPortals($portalRefs);
+        $this->setPortalInformations($portalInformationRefs);
+    }
+}
+
+
+';
+
+        $this->assertEquals($expected, $string);
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]:
```suggestion
        $this->assertSame($expected, $string);
```

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]:
```suggestion
        $this->assertSame($expected, $string);
```

**Comment by alexander-schranz** [comment]
Did quickly replace the assertEquals with assertSame as we are trying to use typesafe phpunit methods where possible.

**Comment by mamazu** [comment]
Yeah, old habits hardly die with the old method. :D

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+use Sulu\Component\Webspace\Manager\WebspaceCollection;
+
+class PhpWebspaceCollectionDumperTest extends TestCase
+{
+    use ProphecyTrait;
+
+    /** @var WebspaceCollection|MockObject */
+    private $webspaceCollection;
+
+    private PhpWebspaceCollectionDumper $webspaceDumper;
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]: looks like php 7.2 does not like this:

```suggestion
     /**
      * @var PhpWebspaceCollectionDumper
      */
    private $webspaceDumper;
```

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+use Prophecy\PhpUnit\ProphecyTrait;
+use Sulu\Component\Webspace\Manager\Dumper\PhpWebspaceCollectionDumper;
+use Sulu\Component\Webspace\Manager\WebspaceCollection;
+
+class PhpWebspaceCollectionDumperTest extends TestCase
+{
+    use ProphecyTrait;
+
+    /** @var WebspaceCollection|MockObject */
+    private $webspaceCollection;
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]: Should be:

```suggestion
    /**
     * @var ObjectProphecy<WebspaceCollection>
     */
    private $webspaceCollection;
```

ObjectProphecy need to be imported.

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+namespace Sulu\Component\Webspace\Tests\Unit\Manager;
+
+use PHPUnit\Framework\TestCase;
+use Prophecy\PhpUnit\ProphecyTrait;
+use Sulu\Component\Webspace\Manager\Dumper\PhpWebspaceCollectionDumper;
+use Sulu\Component\Webspace\Manager\WebspaceCollection;
+
+class PhpWebspaceCollectionDumperTest extends TestCase
+{
+    use ProphecyTrait;
```
Comment (alexander-schranz) [inline src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]: Think this is not required on the 2.4 branch / it does not exist on it. I will add it when merging it back to 2.5.

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
CI seems to fail I added comments to the line failing:

File: src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php (hunk 6 above)
Comment (mamazu) [reply src/Sulu/Component/Webspace/Tests/Unit/Manager/PhpWebspaceCollectionDumperTest.php]: Good old php7.2 :D

**Comment by alexander-schranz** [comment]
@mamazu Thank you!

**Pull request merged**